subsys/bluetooth/host/hci_core.c

```diff
@@ -4851,6 +4851,41 @@ void bt_id_get(bt_addr_le_t *addrs, size_t *count)
 	*count = to_copy;
 }
 
+static void id_create(u8_t id, bt_addr_le_t *addr, u8_t *irk)
+{
+	if (addr && bt_addr_le_cmp(addr, BT_ADDR_LE_ANY)) {
+		bt_addr_le_copy(&bt_dev.id_addr[id], addr);
+	} else {
+		bt_addr_le_create_static(&bt_dev.id_addr[id]);
+		if (addr) {
+			bt_addr_le_copy(addr, &bt_dev.id_addr[id]);
+		}
+	}
+
+#if defined(CONFIG_BT_PRIVACY)
+	if (atomic_test_bit(bt_dev.flags, BT_DEV_READY)) {
+		u8_t zero_irk[16] = { 0 };
+
+		if (irk && memcmp(irk, zero_irk, 16)) {
+			memcpy(&bt_dev.irk[id], irk, 16);
+		} else {
+			bt_rand(&bt_dev.irk[id], 16);
+			if (irk) {
+				memcpy(irk, &bt_dev.irk[id], 16);
+			}
+		}
+	}
+#endif
+	/* Only store if stack was already initialized. Before initialization
+	 * we don't know the flash content, so it's potentially harmful to
+	 * try to write anything there.
+	 */
+	if (IS_ENABLED(CONFIG_BT_SETTINGS) &&
+	    atomic_test_bit(bt_dev.flags, BT_DEV_READY)) {
+		bt_settings_save_id();
+	}
+}
+
 int bt_id_create(bt_addr_le_t *addr, u8_t *irk)
```

Comment:
It doesn't look like you are looking up for duplicated addresses, in case the application already assigned the same address to another instance, I guess that should be considered an invalid use of the API and return an error if that happens.

Reply:
I thought about this, but I'm not sure it's worth increasing the code size for what would essentially be an application bug. The identities are identified by the array index, so providing the same address would not confuse the stack in any way (it may confuse other devices though since e.g. pairing keys would only be available on one of the same-address identities)

Reply:
@Vudentz after thinking even more about this, I went and introduced duplicate identity address guards. We need to check for these either way when generating new random addresses (the chance for a collision is small but still exists).